### PR TITLE
feature(deep link) - add support for merging a received document with the existing doc

### PR DESCRIPTION
**Pull request by hgray-instawork**
As the first step in deep link support, this update supports full merging of the existing dom with an incoming deep link document. The structure of the incoming document remains the same as any other \<navigator\> document structure with the following exceptions:
- For any \<navigator\> node, the addition of an attribute `merge="true"` will enable merging the navigator children. Omitting the attribute will cause any \<navigator\> to replace an existing one of the same name
- Merging will add any child \<nav-route\> nodes which don't exist and will update those which do
- All `selected` values from the current document are reset to "false" to allow the incoming document to set a new selected path

In the example below, the deep link provides the following changes:
1. The shifts-route receives a sub-navigator
2. The tabs-navigator receives an additional 'settings' tab
3. The root-navigator receives a new 'account-popup' route
4. The initial route is now `root-navigator->tabs-route->tabs-navigator->shifts->shifts-route->past-shifts`

Because all incoming data indicated `merge="true"`, all of the changes were additive.

Original:
```xml
<doc xmlns="https://hyperview.org/hyperview">
  <navigator id="root-navigator" type="stack">
    <nav-route id="tabs-route">
      <navigator id="tabs-navigator" type="tab">
        <nav-route id="live-shifts-route" href="/biz-app-hub" selected="true" />
        <nav-route id="shifts-route" href="/biz-app-shift-group-list" />
        <nav-route id="account-route" href="/biz-app-account" />
      </navigator>
    </nav-route>
  </navigator>
</doc>
```

Deep link:
```xml
<doc xmlns="https://hyperview.org/hyperview">
  <navigator id="root-navigator" type="stack" merge="true">
    <nav-route id="tabs-route">
      <navigator id="tabs-navigator" type="tab" merge="true">
        <nav-route id="shifts-route" href="" selected="true">
          <navigator id="shifts" type="tab" merge="true">
            <nav-route id="upcoming-shifts" href="biz-app-shift-group-list" selected="false" />
            <nav-route id="past-shifts" href="/biz-app-shift-group-list'" selected="true" />
          </navigator>
        </nav-route>
        <nav-route id="settings" href="/biz-app-account" selected="false" />
      </navigator>
    </nav-route>
    <nav-route id="account-popup" href="/biz-app-account" modal="true" />
  </navigator>
</doc>
```

Merged:
```xml
<doc xmlns="https://hyperview.org/hyperview">
  <navigator id="root-navigator" type="stack" merge="true">
    <nav-route id="tabs-route" selected="false">
      <navigator id="tabs-navigator" type="tab" merge="true">
        <nav-route id="live-shifts-route" href="/biz_app/gigs/live_shifts" selected="false" />
        <nav-route id="shifts-route" href="/biz_app/gigs/groups" selected="true">
          <navigator id="shifts" type="tab" merge="true">
            <nav-route id="upcoming-shifts" href="/biz_app/gigs/groups" selected="false" />
            <nav-route id="past-shifts" href="/biz_app/gigs/groups" selected="true" />
          </navigator>
        </nav-route>
        <nav-route id="account-route" href="/biz_app/account" selected="false" />
        <nav-route id="settings" href="/biz_app/account" selected="false" />
      </navigator>
    </nav-route>
    <nav-route id="account-popup" href="/biz_app/account" modal="true" />
  </navigator>
</doc>
```

### CODE REVIEW

File: src/services/navigator/helpers.test.ts [hunk 1]
```diff
@@ -47,6 +48,67 @@ const screenDocSource =
...
 
+/**
+ * Test merge original document
+ */
+const mergeOriginalDoc = `
+<doc xmlns="https://hyperview.org/hyperview">
+  <navigator id="root-navigator" type="stack">
+    <nav-route id="tabs-route">
+      <navigator id="tabs-navigator" type="tab">
+        <nav-route id="live-shifts-route" href="{% url 'biz-app-hub' %}" initial="false"/>
```
Comment (flochtililoch) [inline src/services/navigator/helpers.test.ts:59]: Could you remove any Django template tags in the mocked XML? (anything that starts with `{%`

File: src/services/navigator/helpers.ts [hunk 2]
```diff
@@ -301,3 +302,132 @@ export const buildRequest = (
...
   return [navigation, lastPathId || routeId, params];
 };
+
+/**
+ * Create a map of <id, element> from a list of nodes
+ */
+const nodesToMap = (
+  nodes: TypesLegacy.NodeList<TypesLegacy.Node>,
+): Map<string, TypesLegacy.Node> => {
+  const map: Map<string, TypesLegacy.Node> = new Map();
```
Comment (flochtililoch) [inline src/services/navigator/helpers.ts:312]: does using `Map` has any advantage in this situation over a simple object?

File: src/services/navigator/helpers.ts [hunk 3]
```diff
@@ -301,3 +302,132 @@ export const buildRequest = (
...
+): void => {
+  if (!current || !current.childNodes || !newNodes || newNodes.length === 0) {
+    return;
+  }
+
+  // Clean out current node attributes for 'merge' and 'initial'
+  Array.from(current.childNodes).forEach(node => {
+    const element = node as TypesLegacy.Element;
+    if (isNavigationElement(element)) {
+      if (element.localName === TypesLegacy.LOCAL_NAME.NAVIGATOR) {
```
Comment (flochtililoch) [inline src/services/navigator/helpers.ts:352]: When doing the element type check, I think we should also compare `namespaceURI` prop?

File: src/services/navigator/helpers.ts [hunk 4]
```diff
@@ -301,3 +302,132 @@ export const buildRequest = (
...
+    if (node.nodeType === TypesLegacy.NODE_TYPE.ELEMENT_NODE) {
+      const newElement = node as TypesLegacy.Element;
+      if (isNavigationElement(newElement)) {
+        const id = newElement.getAttribute('id');
+        if (id) {
+          const currentElement = currentMap.get(id) as TypesLegacy.Element;
+          if (currentElement) {
+            if (newElement.localName === TypesLegacy.LOCAL_NAME.NAVIGATOR) {
+              const isMergeable =
+                newElement.getAttribute('merge')?.toLowerCase() === 'true';
```
Comment (flochtililoch) [inline src/services/navigator/helpers.ts:372]: Minor, but I think accepted values for boolean attributes are `true`, `false`, `1`, `0`. We shouldn't need to do the case conversion. We don't explicitly support `1` and `0` anywhere I believe.

File: src/services/navigator/helpers.ts (hunk 2 above)
Comment (hgray-instawork) [reply src/services/navigator/helpers.ts:312]: In other languages, the keys are a hash table which provides quick access by key. Maybe not necessary here. Let me know if I should change it over.

File: src/services/navigator/helpers.ts (hunk 3 above)
Comment (hgray-instawork) [reply src/services/navigator/helpers.ts:352]: New code coming

File: src/services/navigator/helpers.ts (hunk 4 above)
Comment (hgray-instawork) [reply src/services/navigator/helpers.ts]: Removed in a few places

File: src/services/navigator/helpers.test.ts (hunk 1 above)
Comment (hgray-instawork) [reply src/services/navigator/helpers.test.ts]: Done

File: src/services/navigator/helpers.ts (hunk 2 above)
Comment (hgray-instawork) [reply src/services/navigator/helpers.ts]: Swapped for obj

File: src/services/navigator/helpers.ts (hunk 3 above)
Comment (hgray-instawork) [reply src/services/navigator/helpers.ts:361]: Done (in the `isNavigationElement` to be more complete)

**Review by flochtililoch — APPROVED** [review]

**Pull request merged**